File: basedatabindingcomponents/src/main/java/com/tanveershafeeprottoy/basedatabindingcomponents/adapters/BaseListAdapter.java
package com.tanveershafeeprottoy.basedatabindingcomponents.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.tanveershafeeprottoy.basedatabindingcomponents.listeners.ListItemOnClickListener;
import com.tanveershafeeprottoy.basedatabindingcomponents.viewholders.BaseRecyclerViewHolder;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

/**
 * @author Tanveer Shafee Prottoy
 */
public class BaseListAdapter<T> extends RecyclerView.Adapter<BaseRecyclerViewHolder> {
    private int resourceId;
    private int variableId;
    private ListItemOnClickListener listItemOnClickListener;
    private List<T> objList;

    public BaseListAdapter(int resourceId, int variableId,
                           ListItemOnClickListener listItemOnClickListener) {
        this.resourceId = resourceId;
        this.variableId = variableId;
        this.listItemOnClickListener = listItemOnClickListener;
    }

    @Override
    @NonNull
    public BaseRecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new BaseRecyclerViewHolder<T>(DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), resourceId, parent, false), variableId, listItemOnClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseRecyclerViewHolder holder, int position) {
        try {
            holder.bind(objList.get(position));
        }
        catch(Exception e) {

        }
    }

    @Override
    public int getItemCount() {
        try {
            return objList.size();
        }
        catch(NullPointerException n) {
            return 0;
        }
    }

    public void setData(List<T> objList) {
        this.objList = objList;
        notifyDataSetChanged();
    }
}

File: basedatabindingcomponents/src/main/java/com/tanveershafeeprottoy/basedatabindingcomponents/adapters/BaseListAdapter.kt
package com.tanveershafeeprottoy.basedatabindingcomponents.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.tanveershafeeprottoy.basedatabindingcomponents.listeners.ListItemOnClickListener
import com.tanveershafeeprottoy.basedatabindingcomponents.viewholders.BaseRecyclerViewHolder

open class BaseListAdapter<T>(private val resourceId: Int, private val variableId: Int,
                              private val listItemOnClickListener:
                              ListItemOnClickListener?) : RecyclerView.Adapter<BaseRecyclerViewHolder>() {
    private var objList: MutableList<T>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecyclerViewHolder {
        return BaseRecyclerViewHolder(DataBindingUtil.inflate(LayoutInflater.from(parent.context), resourceId, parent, false), variableId, listItemOnClickListener)
    }

    override fun onBindViewHolder(holder: BaseRecyclerViewHolder, position: Int) {
        try {
            holder.bind(objList!![position])
        }
        catch(e: Exception) {

        }
    }

    override fun getItemCount(): Int {
        return try {
            objList!!.size
        }
        catch(n: NullPointerException) {
            0
        }
    }

    fun setData(objList: MutableList<T>?) {
        this.objList = objList
        notifyDataSetChanged()
    }
}